flink-runtime/src/main/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorContext.java

```diff
@@ -431,4 +489,125 @@ private <V> V callInCoordinatorThread(Callable<V> callable, String errorMessage)
             throw new FlinkRuntimeException(errorMessage, t);
         }
     }
+
+    private void assignSplitsToAttempts(SplitsAssignment<SplitT> assignment) {
+        assignment.assignment().forEach((index, splits) -> assignSplitsToAttempts(index, splits));
+    }
+
+    private void assignSplitsToAttempts(int subtaskIndex, List<SplitT> splits) {
+        getRegisteredAttempts(subtaskIndex)
+                .forEach(attempt -> assignSplitsToAttempt(subtaskIndex, attempt, splits));
+    }
+
+    private void assignSplitsToAttempt(int subtaskIndex, int attemptNumber, List<SplitT> splits) {
+        checkAttemptReaderReady(subtaskIndex, attemptNumber);
+
+        final AddSplitEvent<SplitT> addSplitEvent;
+        try {
+            addSplitEvent = new AddSplitEvent<>(splits, splitSerializer);
```

Comment:
It's better not send this message if splits is empty or null.

Reply:
fixed.